src/fluentd/src/fluent-plugin-pgjson/lib/fluent/plugin/out_pgjson.rb

```diff
@@ -140,8 +140,65 @@ def format(tag, time, record)
       [Time.at(time).strftime(@time_format), record].to_msgpack
     end
 
+    def insert_framework(hex_id, time, record)
+      # This function try to insert the framework snapshot into framework history table.
+      # In some cases, the framework controller may have duplicate logs about one framework attempt,
+      # or there has been already successful inserted record before.
+      # To handle it, `insert_framework` executes a "SELECT" first.
+      # If the uid exists in the table, it will ignore it safely.
+      # Any error should be raised.
+      thread = Thread.current
+      frameworkName = record["objectSnapshot"]["metadata"]["name"]
+      attemptIndex = record["objectSnapshot"]["status"]["attemptStatus"]["id"]
+      historyType = "retry"
+      snapshot = record_value(record["objectSnapshot"])
+      # use frameworkName + attemptIndex + historyType to generate a uid
+      uid = Digest::MD5.hexdigest "#{frameworkName}+#{attemptIndex}+#{historyType}"
+      # select from framework_history, ensure there is no corresponding history object
+      selectResult = thread[:conn].exec_params('SELECT uid from framework_history where uid=$1', [uid])
+      if selectResult.cmd_tuples == 0
+          # if there is no existing records, try to insert a new one.
+          thread[:conn].exec_params("INSERT INTO framework_history (\"#{@insertedAt_col}\", \"#{@updatedAt_col}\", \"#{@uid_col}\", \"#{@frameworkName_col}\", \"#{@attemptIndex_col}\", \"#{@historyType_col}\", \"#{@snapshot_col}\") " +
+            "VALUES ($1, $2, $3, $4, $5, $6, $7)", [time, time, uid, frameworkName, attemptIndex, historyType, snapshot]
+          )
+      else
+        # if there is an existing record, ignore it.
+        log.warn "[pgjson] chunk #{hex_id}: ignored framework snapshot object, uid=#{uid}"
```

Comment:
log.warn "[pgjson] chunk #{hex_id}: ignored framework snapshot object as it already exist, uid=#{uid}"

Reply:
Done